apimachinery/src/meta/mod.rs

```diff
@@ -268,3 +286,334 @@ fn gr_fromstr() {
+                Token::BorrowedStr("v1alpha1"),
+                Token::Str("kind"),
+                Token::Some,
+                Token::BorrowedStr("Test"),
+                Token::StructEnd,
+            ],
+        );
+
+        // Reversed order of fields
+        assert_de_tokens(
+            &t,
+            &[
+                Token::Struct {
+                    name: "TypeMeta",
+                    len: 2,
+                },
+                Token::Str("kind"),
+                Token::Some,
+                Token::BorrowedStr("Test"),
+                Token::Str("apiVersion"),
+                Token::Some,
+                Token::BorrowedStr("v1alpha1"),
+                Token::StructEnd,
+            ],
+        );
+
+        // No apiVersion/kind is also ok
+        assert_de_tokens(
+            &t,
+            &[
+                Token::Struct {
+                    name: "TypeMeta",
+                    len: 0,
+                },
+                Token::StructEnd,
+            ],
+        );
+    }
+
+    #[test]
+    fn test_typemeta_serde_error() {
+        assert_de_tokens_error::<TypeMetaImpl<TestType>>(
+            &[
+                Token::Struct {
+                    name: "TypeMeta",
+                    len: 1,
+                },
+                Token::Str("kind"),
+                Token::Some,
+                Token::BorrowedStr("TestType"),
+                Token::StructEnd,
+            ],
+            "missing field `apiVersion`",
+        );
+
+        assert_de_tokens_error::<TypeMetaImpl<TestType>>(
+            &[
+                Token::Struct {
+                    name: "TypeMeta",
+                    len: 1,
+                },
+                Token::Str("apiVersion"),
+                Token::Some,
+                Token::BorrowedStr("bogus"),
+                Token::StructEnd,
+            ],
+            "missing field `kind`",
+        );
+
+        assert_de_tokens_error::<TypeMetaImpl<TestType>>(
+            &[
+                Token::Struct {
+                    name: "TypeMeta",
+                    len: 1,
+                },
+                Token::Str("apiVersion"),
+                Token::Some,
+                Token::Str("v1alpha1"),
+                Token::Str("apiVersion"),
+                Token::StructEnd,
+            ],
+            "duplicate field `apiVersion`",
+        );
+
+        assert_de_tokens_error::<TypeMetaImpl<TestType>>(
+            &[
+                Token::Struct {
+                    name: "TypeMeta",
+                    len: 2,
+                },
+                Token::Str("kind"),
+                Token::Some,
+                Token::Str("NotTest"),
+                Token::Str("apiVersion"),
+                Token::Some,
+                Token::Str("v1alpha1"),
+                Token::StructEnd,
+            ],
+            "invalid value: v1alpha1/NotTest, expected v1alpha1/Test",
+        );
+    }
+}
+
+pub trait ResourceScope {
```

Comment:
This feels a little awkward, particularly with the impls below, I guess we'll see where its used later...

Reply:
Yes, I'm not completely happy with these "name" bits. Alternative suggestions welcome :-\